R/pipeline.R

```diff
@@ -0,0 +1,224 @@
+#' Run an Rt Estimation Model Pipeline
+#'
+#' This function runs a complete pipeline for fitting an Rt estimation model,
+#' using the `EpiNow2` model, based on a configuration file. The pipeline
+#' processes the model, logs its progress, and handles errors by logging
+#' warnings and setting the pipeline status. Output and logs are written to
+#' the specified directories. Additionally, support for uploading logs and
+#' outputs to a blob storage container is planned.
+#'
+#' @param config_path A string specifying the file path to the JSON
+#' configuration file. The configuration file should include at least `job_id`
+#' and `task_id` to manage the logging and task metadata.
+#' @param blob_storage_container Optional. The name of the blob storage
+#' container to which logs and outputs will be uploaded. If NULL, no upload
+#' will occur. (Planned feature, not currently implemented)
+#' @param output_dir A string specifying the directory where output, logs, and
+#' other pipeline artifacts will be saved. Defaults to the root directory ("/").
+#'
+#' @details
+#' The function reads the configuration from a JSON file and uses this to set
+#' up the job and task identifiers. It creates an output directory structure
+#' based on these IDs and starts logging the process in a file. The main
+#' pipeline process is handled by `process_pipeline()`, with errors caught and
+#' logged as warnings. The function will log the success or failure of the run.
+#'
+#' Logs are written to a file in the output directory, and console output is
+#' also mirrored in this log file. Error handling is in place to capture any
+#' issues during the pipeline execution and ensure they are logged
+#' appropriately.
+#'
+#' During the execution of the pipeline, the following output files are
+#' expected to be generated:
+#'
+#' - **Model Output**: An RDS file of the fitted model is saved in the
+#' task-specific directory (`model.rds`).
+#' - **Samples**: Parquet files containing the model's sample outputs are saved
+#' in a `samples` subdirectory, named using the `task_id` (e.g.,
+#' `task_id.parquet`).
+#' - **Summaries**: Parquet files summarizing the model's results are saved in
+#' a `summaries` subdirectory, also named using the `task_id` (e.g.,
+#' `task_id.parquet`).
+#' - **Logs**: A `logs.txt` file is generated in the task directory, capturing
+#' both console and error messages.
+#'
+#' The output directory structure will follow this format:
+#' ```
+#' <output_dir>/
+#' └── <job_id>/
+#'     ├── samples/
+#'     │   └── <task_id>.parquet
+#'     ├── summaries/
+#'     │   └── <task_id>.parquet
+#'     └── tasks/
+#'         └── <task_id>/
+#'             ├── model.rds
+#'             └── logs.txt
+#' ```
```

Comment:
Can you explain here too -- if you run this in batch but don't specify `blob_storage`, will the outputs be lost?

Reply:
They'll live in a local folder. If you run it in a container without bind mounting the output location to your computer's file system, they will be lost.

I think the end goal will be to bind mount mount the output location to the blob? Or maybe to the node's local storage, and then upload at the end?

Reply:
I'll add more to the docs on Azure handling when I write the Azure bits into the pipeline.